### DISCUSSION TITLE
Doubt about the results in OMEdit for this example?

### DISCUSSION THREAD

**Discussion by Amuri44** [OpenModelica/OpenModelica#8568] [Q&A] [answered]
I have and `add` with its input `parameter input` and I see abnormal behavior?

**Comment by casella** [comment]
This is indeed weird.

However, I wonder why one should ever write
```modelica
  parameter input Real X1 = 1;
```
either `X1` is a parameter, with a binding equation to provide its value, or it is an input, in which case it cannot be a parameter and it shouldn't have a binding equation. It can't be both things at the same time, that doesn't mean anything to me.

@perost, do you think such a construct should actually mean something, according to the specification?

**Reply by Amuri44** [reply]
I see

**Reply by perost** [reply]
As for the specification I don't think there's anything that explicitly forbids `parameter input`, though I don't think it says anything about what it means either.

**Reply by Amuri44** [reply]
@casella @perost , so any proper keywords before `X1, X2` to avoid such these problems?

**Reply by casella** [reply]
I guess `parameter` is what you need. When you simulate in OMEdit it will remain constant, and if you export an FMU, you can still change it at runtime.

**Reply by Amuri44** [reply]
Ok, thank you very much!

**Comment by Arinomo** [comment]
Hi, just to add to the discussion, 
@Amuri44 : if you enable the option `evaluate all parameters` in Simulation Setup --> Translation Flags then the output of `add` will stay at 3 (per OME v1.19.0-dev-590-g0897872ac1 win64). 

@casella : I wasn't aware that the keywords `parameter` and `input` could be used together and so the variable will be defined as `ìnput` in the `modelDescription.xml`, if i recall correctly, this is what `tunable parameter` in the FMU was suppose to do. But if i disabled `evaluate parameter` and re-simulate the model with different X1 and/or X2, the output of `add` still jumps to `6` regardless the input values. So i guess the combination is not 100% legal?


Edit:
when only one variable is defined as `parameter input`, the simulation result (i.e. the output of `add`) will be completly different.
`parameter input Real X1 = 1.0;` and `parameter Real X2 = 2.0;` make `add.y` to increase to `5` 
`parameter Real X1 = 1.0;` and `parameter input Real X2 = 2.0;` make `add.y` to increase to `4`

**Reply by casella** [reply]
As @perost wrote, although the semantics of parameter input is not defined in the Modelica syntax, it is not explicitly forbidden either. Which means, the code is legal Modelica, but the behaviour is undefined. I understand such corner cases often happen in compiler implementations. 

Tunable parameters are only defined in FMI. I honestly find the concept a bit fishy: you write a model in Modelica under the assumption that a certain symbol has parameter variability (i.e. it doesn't change during the simulation), and then you do change it.

In any case, the FMI specification (Sect. 2.2.7) clearly explains that this is conceptually equivalent to stopping the simulation at an event, change the tunable parameters, and then start a new simulation from the same final states of the previous one. This is meta-modelling, that cannot really be defined in Modelica - what the Modelica code defines is one simulation run, including initialization, not multiple linked ones.

I'm not sure what is the expectation of the FMI community about how one could specify a tunable parameter in the original Modelica model. I can ask. But as far as I know, `parameter input` doesn't mean anything according to the current Modelica specification, so it's not too clear to me from where the FMI code generation should infer that a parameter is tunable. @lochel, you may comment on that.

**Comment by casella** [comment, marked as answer]
See [MLS #3140](https://github.com/modelica/ModelicaSpecification/issues/3140) and [MLS #3141](https://github.com/modelica/ModelicaSpecification/issues/3141). Let's continue the discussion there.
